### PR TITLE
CLOUDP-223587: Cleanup orphan GCP private links

### PR DESCRIPTION
**Pull request by josvazg**
✅ [Managed to remove all GCP left-over addresses and forwarding rules pairs](https://github.com/mongodb/mongodb-atlas-kubernetes/actions/runs/8050671222/job/21986791683)

### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

### CODE REVIEW

**Comment by josvazg** [comment]
Screenshot 2024-02-26 at 15 46 36

**Comment by helderjs** [comment]
@josvazg Do you know why these GCP resources are becoming orphans?

**Review by helderjs — APPROVED** [review]
LGTM

**Comment by josvazg** [comment]
> @josvazg Do you know why these GCP resources are becoming orphans?

Not really. I thought we could create a follow up for that. But here I wanted to have a fix that would eliminate the recurrent issue for now.

I suspect they become orphan on some tests failures that do not clean up properly after themselves. I do not believe it happens all the time, it is going to be tricky to debug.

I will create and link the follow up here.

**Comment by josvazg** [comment]
Follow up to investigate how the e2e cleanup might fails and turn those resources into orphans CLOUDP-234041

File: tools/clean/provider/gcp.go [hunk 1]
```diff
@@ -100,6 +103,73 @@ func (gcp *GCP) DeletePrivateEndpoint(ctx context.Context, groupName, sAttachmen
...
+		Project: gcp.projectID,
+		Region:  region,
+	})
+	for {
+		addr, err := addresses.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return err
```
Comment (s-urbaniak) [inline tools/clean/provider/gcp.go:117]: wrap using `fmt.Errorf("error while iterating addresses in project %v region $v: %w", gcp.projectID, region, err)

File: tools/clean/provider/gcp.go [hunk 2]
```diff
@@ -100,6 +103,73 @@ func (gcp *GCP) DeletePrivateEndpoint(ctx context.Context, groupName, sAttachmen
...
+		}
+		suffix := fmt.Sprintf("subnetworks/%s", subnet)
+		if !strings.HasSuffix(addr.GetSubnetwork(), suffix) {
+			fmt.Println(text.FgYellow.Sprintf("\tAddress %s(%s) skipped, not in %s\n",
+				addr.GetName(), addr.GetAddress(), subnet))
+			continue
+		}
+		createdAt, err := asTime(addr.GetCreationTimestamp())
+		if err != nil {
+			return err
```
Comment (s-urbaniak) [inline tools/clean/provider/gcp.go:127]: `return fmt.Errorf("error parsing creation timestamp %q: %w", addr.GetCreationTimestamp(), err)`

File: tools/clean/provider/gcp.go [hunk 3]
```diff
@@ -100,6 +103,73 @@ func (gcp *GCP) DeletePrivateEndpoint(ctx context.Context, groupName, sAttachmen
...
+		}
+		if time.Since(createdAt) < time.Duration(lifetime)*time.Hour {
+			fmt.Println(text.FgYellow.Sprintf(
+				"\tAddress %s(%s) skipped once created less than %d hours ago\n",
+				addr.GetName(), addr.GetAddress(), lifetime))
+		}
+		frName, err := expectForwardingRule(addr.GetUsers())
+		if err != nil {
+			return err
```
Comment (s-urbaniak) [inline tools/clean/provider/gcp.go:136]: this is actually fine not wrapping, as the error is generated here, only once, and is specific :-)

File: tools/clean/provider/gcp.go [hunk 4]
```diff
@@ -100,6 +103,73 @@ func (gcp *GCP) DeletePrivateEndpoint(ctx context.Context, groupName, sAttachmen
...
+		}
+
+		if frName != "" {
+			if err := gcp.deleteForwardingRule(ctx, frName, region); err != nil {
+				return err
```
Comment (s-urbaniak) [inline tools/clean/provider/gcp.go:141]: `return fmt.Errorf("error deleting forwarding rule %q in region %q: %w", region, frName, err)`

File: tools/clean/provider/gcp.go [hunk 5]
```diff
@@ -100,6 +103,73 @@ func (gcp *GCP) DeletePrivateEndpoint(ctx context.Context, groupName, sAttachmen
...
+			}
+			fmt.Println(text.FgGreen.Sprintf("\tDeleted forwarding rule %s for %s\n",
+				frName, addr.GetAddress()))
+		}
+		if err := gcp.deleteIPAddress(ctx, addr.GetName(), region); err != nil {
+			return err
```
Comment (s-urbaniak) [inline tools/clean/provider/gcp.go:147]: `return fmt.Errorf("error deleting IP address %q in region %q: %w", addr.GetName(), frName, err)`

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
just a few error handling nits, else lgtm

File: Makefile [hunk 6]
```diff
@@ -128,15 +128,15 @@ recompute-licenses: ## Recompute the licenses.csv only if needed (gomod was chan
 
 licenses-up-to-date:
 	@if [ "$(GOMOD_SHA)" != "$(GOMOD_LICENSES_SHA)" ]; then \
-	echo "licenses.csv needs ot be recalculated: run 'make licenses.csv'"; exit 1; else\
-	echo "licenses.csv is OK! (up to date)"; fi
+	echo "licenses.csv needs to be recalculated: git rebase AND run 'make licenses.csv'"; exit 1; \
```
Comment (josvazg) [inline Makefile:131]: This tripped me for quite a while.

The check will fail if you need to rebase as well, even thought locally the `licenses.csv` seems up to date, if the diff includes a rebase, the go.mod will be different and thus the `licenses.csv` need to be recomputed.

While I was trying to fix this I also removed our dependency on shasum and use git SHAs directly instead.

**Review by helderjs — APPROVED** [review]

**Review by s-urbaniak — APPROVED** [review]
lgtm

**Pull request merged**